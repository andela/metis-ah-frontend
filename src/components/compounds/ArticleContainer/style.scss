.article-inner-container {
	padding: 55px 10px;
	flex: 1;
	display: flex;
	flex-direction: column;
	.article-loader {
		display: flex;
		flex:1;
		width: 100%;
    height: 64px;
    justify-content: center;
    align-items: center
	}
	.article-loader:after {
		content: ' ';
		display: block;
		width: 56px;
		height: 56px;
		margin: 1px;
		border-radius: 50%;
		border: 5px solid #13b47a;
		border-color: #13b47a transparent #13b47a transparent;
		animation: article-loader 1.2s linear infinite;
	}

	.article-container-card {
    background-color: white;
		border-radius: 5px;
		display: flex;
		flex-direction: column;
		width: 380px;
		padding: 1%;
		margin: auto;
		align-items: center;
		box-shadow: 0 2px 4px rgba(0,0,0,0.11), 0 2px 4px rgba(0,0,0,0.18);
		.notification-icon {
			font-size: 30px;
			color: #13b47a;
		}
		p {
			font-size: 20px;
			padding-top: 15px;
		}
  }

	@keyframes article-loader {
		0% {
			transform: rotate(0deg);
		}
		100% {
			transform: rotate(360deg);
		}
	}
}

@media screen and (max-width: 425px) {
	.article-inner-container {
		.article-container-card {
			width: 280px;
			.notification-icon {
				font-size: 25px;
			}
			p {
				font-size: 17px;
			}
		}
	}
}

@media screen and (max-width: 768px) {
	.article-inner-container {
		padding: 35px 15px;
	}
}
