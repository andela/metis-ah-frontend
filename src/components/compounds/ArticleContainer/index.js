
import React, { Component } from 'react';
import { withRouter, Link } from 'react-router-dom';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import ArticleCard from 'Components/atoms/ArticleCard';
import { getArticle } from 'Actions/article';
import sortArrayContent from '../../util/helper';
import './style.scss';

class ArticleContainer extends Component {
  constructor(props) {
    super(props);
    const {
      match: { params: category }
    } = this.props;
    this.state = {
      category
    };
  }

  componentDidMount() {
    const { getArticles, match, categories } = this.props;
    const heroContent = sortArrayContent(categories, match.params.category);
    getArticles(match.params.category, heroContent);
  }

  static getDerivedStateFromProps(nextProps, prevState) {
    const {
      match: {
        params: { category }
      }
    } = nextProps;
    const { categories } = nextProps;
    const heroContent = sortArrayContent(categories, category);
    if (prevState.category.category !== nextProps.match.params.category) {
      nextProps.getArticles(category, heroContent);
      return {
        category: { category }
      };
    }
    return null;
  }

  render() {
    const { articles, loading } = this.props;

    const mappedArticleCards = articles.map((article) => {
      const {
        id,
        imageUrl,
        slug,
        title,
        description,
        createdAt,
        User: { firstname, lastname }
      } = article;
      return (
        <ArticleCard
          key={slug}
          id={id}
          img={imageUrl}
          title={title}
          description={description}
          name={`${firstname} ${lastname}`}
          time={createdAt}
        />
      );
    });
    return (
      <div className="container article-inner-container">
        {mappedArticleCards}
        {articles.length < 1 && (loading && <div className="article-loader" />)}
        {articles.length < 1
          && (!loading && (
            <div className="article-container-card">
              <i className="fas fa-exclamation notification-icon" />
              <p>
                Oops!, No Article currently exits in this category.
                <br />
                We would like you to be the first to write in this category
              </p>
              <br />
              <Link to="/articles/new">WRITE</Link>
            </div>
          ))}
      </div>
    );
  }
}

const mapStateToProps = state => ({
  loading: state.article.loading,
  articles: state.article.articles,
  categories: state.categories.categories
});

const mapDispatchToProps = dispatch => ({
  getArticles: (categoryId, heroContent) => dispatch(getArticle(categoryId, heroContent))
});

ArticleContainer.propTypes = {
  articles: PropTypes.arrayOf(PropTypes.object),
  loading: PropTypes.bool.isRequired,
  match: PropTypes.objectOf(PropTypes.any).isRequired,
  getArticles: PropTypes.func.isRequired,
  categories: PropTypes.arrayOf(PropTypes.object).isRequired
};

ArticleContainer.defaultProps = {
  articles: []
};
export default withRouter(
  connect(
    mapStateToProps,
    mapDispatchToProps
  )(ArticleContainer)
);
