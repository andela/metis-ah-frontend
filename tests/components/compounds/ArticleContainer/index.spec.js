import React from 'react';
import { BrowserRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import { mount } from 'enzyme';
import thunk from 'redux-thunk';
import configureStore from 'redux-mock-store';
import ArticleContainer from 'Components/compounds/ArticleContainer';
import initialState from '../../../__mocks__/articleContainerMock';

const middleware = [thunk];
const mockStore = configureStore(middleware);
let wrapper;
let store;
let timeFormat;

describe('article container component test', () => {
  beforeEach(() => {
    timeFormat = jest.fn();
    store = mockStore(initialState);
    wrapper = mount(
      <Provider store={store}>
        <BrowserRouter>
          <ArticleContainer timeFormat={timeFormat} />
        </BrowserRouter>
      </Provider>
    );
  });

  it('should render the article container correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
