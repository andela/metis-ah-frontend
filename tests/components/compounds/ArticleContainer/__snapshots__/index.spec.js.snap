// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`article container component test should render the article container correctly 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <BrowserRouter>
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <withRouter(Connect(ArticleContainer))
        timeFormat={[MockFunction]}
      >
        <Route>
          <Connect(ArticleContainer)
            history={
              Object {
                "action": "POP",
                "block": [Function],
                "createHref": [Function],
                "go": [Function],
                "goBack": [Function],
                "goForward": [Function],
                "length": 1,
                "listen": [Function],
                "location": Object {
                  "hash": "",
                  "pathname": "/",
                  "search": "",
                  "state": undefined,
                },
                "push": [Function],
                "replace": [Function],
              }
            }
            location={
              Object {
                "hash": "",
                "pathname": "/",
                "search": "",
                "state": undefined,
              }
            }
            match={
              Object {
                "isExact": true,
                "params": Object {},
                "path": "/",
                "url": "/",
              }
            }
            timeFormat={[MockFunction]}
          >
            <ArticleContainer
              articles={
                Array [
                  Object {
                    "User": Object {
                      "firstname": "Tomi",
                      "lastname": "Adebanjo",
                    },
                    "createdAt": "23-09-2018",
                    "description": "We can't Keep blaming behavior on the robots",
                    "imageUrl": "rectangle1",
                    "slug": "asasa",
                    "title": "The Internet's problems Can't Be solved with an Algorithm",
                  },
                  Object {
                    "User": Object {
                      "firstname": "Tomi",
                      "lastname": "Adebanjo",
                    },
                    "createdAt": "23-09-2018",
                    "description": "We can't Keep blaming behavior on the robots",
                    "imageUrl": "rectangle2",
                    "slug": "asasaa",
                    "title": "The Internet's problems Can't Be solved with an Algorithm",
                  },
                ]
              }
              categories={
                Array [
                  Object {
                    "description": "It has become appallingly obvious that our technology has exceeded our humanity.",
                    "id": 1,
                    "name": "FASHION",
                    "poster": "https://res.cloudinary.com/dbsxxymfz/image/upload/v1539947552/eroooo_1.png",
                  },
                ]
              }
              getArticles={[Function]}
              history={
                Object {
                  "action": "POP",
                  "block": [Function],
                  "createHref": [Function],
                  "go": [Function],
                  "goBack": [Function],
                  "goForward": [Function],
                  "length": 1,
                  "listen": [Function],
                  "location": Object {
                    "hash": "",
                    "pathname": "/",
                    "search": "",
                    "state": undefined,
                  },
                  "push": [Function],
                  "replace": [Function],
                }
              }
              loading={false}
              location={
                Object {
                  "hash": "",
                  "pathname": "/",
                  "search": "",
                  "state": undefined,
                }
              }
              match={
                Object {
                  "isExact": true,
                  "params": Object {},
                  "path": "/",
                  "url": "/",
                }
              }
              metadata={Object {}}
              timeFormat={[MockFunction]}
            >
              <div
                className="container article-inner-container"
              >
                <withRouter(ArticleCard)
                  description="We can't Keep blaming behavior on the robots"
                  img="rectangle1"
                  key="asasa"
                  name="Tomi Adebanjo"
                  time="23-09-2018"
                  title="The Internet's problems Can't Be solved with an Algorithm"
                >
                  <Route>
                    <ArticleCard
                      description="We can't Keep blaming behavior on the robots"
                      history={
                        Object {
                          "action": "POP",
                          "block": [Function],
                          "createHref": [Function],
                          "go": [Function],
                          "goBack": [Function],
                          "goForward": [Function],
                          "length": 1,
                          "listen": [Function],
                          "location": Object {
                            "hash": "",
                            "pathname": "/",
                            "search": "",
                            "state": undefined,
                          },
                          "push": [Function],
                          "replace": [Function],
                        }
                      }
                      id={0}
                      img="rectangle1"
                      location={
                        Object {
                          "hash": "",
                          "pathname": "/",
                          "search": "",
                          "state": undefined,
                        }
                      }
                      match={
                        Object {
                          "isExact": true,
                          "params": Object {},
                          "path": "/",
                          "url": "/",
                        }
                      }
                      name="Tomi Adebanjo"
                      time="23-09-2018"
                      title="The Internet's problems Can't Be solved with an Algorithm"
                    >
                      <div
                        className="articleCard-container"
                        onClick={[Function]}
                      >
                        <div
                          className="columns articleCard"
                        >
                          <div
                            className="column is-one-fifth padding__0"
                          >
                            <img
                              alt=""
                              className="articleCard__img"
                              src="rectangle1"
                            />
                          </div>
                          <div
                            className="column padding__0 articleCard__content"
                          >
                            <h2
                              className="articleCard__title"
                            >
                              The Internet's problems Can't Be solved with an Algorithm
                            </h2>
                            <p
                              className="articleCard__subtitle"
                            >
                              We can't Keep blaming behavior on the robots
                            </p>
                            <p
                              className="articleCard__name"
                            >
                              Tomi Adebanjo
                            </p>
                            <p
                              className="articleCard__time"
                            >
                              just now
                            </p>
                          </div>
                        </div>
                      </div>
                    </ArticleCard>
                  </Route>
                </withRouter(ArticleCard)>
                <withRouter(ArticleCard)
                  description="We can't Keep blaming behavior on the robots"
                  img="rectangle2"
                  key="asasaa"
                  name="Tomi Adebanjo"
                  time="23-09-2018"
                  title="The Internet's problems Can't Be solved with an Algorithm"
                >
                  <Route>
                    <ArticleCard
                      description="We can't Keep blaming behavior on the robots"
                      history={
                        Object {
                          "action": "POP",
                          "block": [Function],
                          "createHref": [Function],
                          "go": [Function],
                          "goBack": [Function],
                          "goForward": [Function],
                          "length": 1,
                          "listen": [Function],
                          "location": Object {
                            "hash": "",
                            "pathname": "/",
                            "search": "",
                            "state": undefined,
                          },
                          "push": [Function],
                          "replace": [Function],
                        }
                      }
                      id={0}
                      img="rectangle2"
                      location={
                        Object {
                          "hash": "",
                          "pathname": "/",
                          "search": "",
                          "state": undefined,
                        }
                      }
                      match={
                        Object {
                          "isExact": true,
                          "params": Object {},
                          "path": "/",
                          "url": "/",
                        }
                      }
                      name="Tomi Adebanjo"
                      time="23-09-2018"
                      title="The Internet's problems Can't Be solved with an Algorithm"
                    >
                      <div
                        className="articleCard-container"
                        onClick={[Function]}
                      >
                        <div
                          className="columns articleCard"
                        >
                          <div
                            className="column is-one-fifth padding__0"
                          >
                            <img
                              alt=""
                              className="articleCard__img"
                              src="rectangle2"
                            />
                          </div>
                          <div
                            className="column padding__0 articleCard__content"
                          >
                            <h2
                              className="articleCard__title"
                            >
                              The Internet's problems Can't Be solved with an Algorithm
                            </h2>
                            <p
                              className="articleCard__subtitle"
                            >
                              We can't Keep blaming behavior on the robots
                            </p>
                            <p
                              className="articleCard__name"
                            >
                              Tomi Adebanjo
                            </p>
                            <p
                              className="articleCard__time"
                            >
                              just now
                            </p>
                          </div>
                        </div>
                      </div>
                    </ArticleCard>
                  </Route>
                </withRouter(ArticleCard)>
                <PaginationBoxView
                  activeClassName="selected"
                  activeLinkClassName="activePage"
                  breakLabel="..."
                  containerClassName="pagination"
                  disableInitialCallback={false}
                  disabledClassName="disabled"
                  marginPagesDisplayed={3}
                  nextClassName="next"
                  nextLabel={
                    <span>
                      Next
                    </span>
                  }
                  nextLinkClassName="next-icon"
                  onPageChange={[Function]}
                  pageCount={10}
                  pageRangeDisplayed={5}
                  previousClassName="previous"
                  previousLabel={
                    <span>
                      Prev
                    </span>
                  }
                  previousLinkClassName="prev-icon"
                >
                  <ul
                    className="pagination"
                  >
                    <li
                      className="previous disabled"
                    >
                      <a
                        className="prev-icon"
                        onClick={[Function]}
                        onKeyPress={[Function]}
                        role="button"
                        tabIndex="0"
                      >
                        <span>
                          Prev
                        </span>
                      </a>
                    </li>
                    <PageView
                      activeClassName="selected"
                      activeLinkClassName="activePage"
                      key="0"
                      onClick={[Function]}
                      page={1}
                      selected={true}
                    >
                      <li
                        className="selected"
                      >
                        <a
                          aria-current="page"
                          aria-label="Page 1 is your current page"
                          className="activePage"
                          onClick={[Function]}
                          onKeyPress={[Function]}
                          role="button"
                          tabIndex="0"
                        >
                          1
                        </a>
                      </li>
                    </PageView>
                    <PageView
                      activeClassName="selected"
                      activeLinkClassName="activePage"
                      key="1"
                      onClick={[Function]}
                      page={2}
                      selected={false}
                    >
                      <li>
                        <a
                          aria-current={null}
                          aria-label="Page 2"
                          onClick={[Function]}
                          onKeyPress={[Function]}
                          role="button"
                          tabIndex="0"
                        >
                          2
                        </a>
                      </li>
                    </PageView>
                    <PageView
                      activeClassName="selected"
                      activeLinkClassName="activePage"
                      key="2"
                      onClick={[Function]}
                      page={3}
                      selected={false}
                    >
                      <li>
                        <a
                          aria-current={null}
                          aria-label="Page 3"
                          onClick={[Function]}
                          onKeyPress={[Function]}
                          role="button"
                          tabIndex="0"
                        >
                          3
                        </a>
                      </li>
                    </PageView>
                    <PageView
                      activeClassName="selected"
                      activeLinkClassName="activePage"
                      key="3"
                      onClick={[Function]}
                      page={4}
                      selected={false}
                    >
                      <li>
                        <a
                          aria-current={null}
                          aria-label="Page 4"
                          onClick={[Function]}
                          onKeyPress={[Function]}
                          role="button"
                          tabIndex="0"
                        >
                          4
                        </a>
                      </li>
                    </PageView>
                    <PageView
                      activeClassName="selected"
                      activeLinkClassName="activePage"
                      key="4"
                      onClick={[Function]}
                      page={5}
                      selected={false}
                    >
                      <li>
                        <a
                          aria-current={null}
                          aria-label="Page 5"
                          onClick={[Function]}
                          onKeyPress={[Function]}
                          role="button"
                          tabIndex="0"
                        >
                          5
                        </a>
                      </li>
                    </PageView>
                    <PageView
                      activeClassName="selected"
                      activeLinkClassName="activePage"
                      key="5"
                      onClick={[Function]}
                      page={6}
                      selected={false}
                    >
                      <li>
                        <a
                          aria-current={null}
                          aria-label="Page 6"
                          onClick={[Function]}
                          onKeyPress={[Function]}
                          role="button"
                          tabIndex="0"
                        >
                          6
                        </a>
                      </li>
                    </PageView>
                    <BreakView
                      breakLabel="..."
                      key="6"
                      onClick={[Function]}
                    >
                      <li
                        className="break"
                      >
                        <a
                          onClick={[Function]}
                          onKeyPress={[Function]}
                          role="button"
                          tabIndex="0"
                        >
                          ...
                        </a>
                      </li>
                    </BreakView>
                    <PageView
                      activeClassName="selected"
                      activeLinkClassName="activePage"
                      key="7"
                      onClick={[Function]}
                      page={8}
                      selected={false}
                    >
                      <li>
                        <a
                          aria-current={null}
                          aria-label="Page 8"
                          onClick={[Function]}
                          onKeyPress={[Function]}
                          role="button"
                          tabIndex="0"
                        >
                          8
                        </a>
                      </li>
                    </PageView>
                    <PageView
                      activeClassName="selected"
                      activeLinkClassName="activePage"
                      key="8"
                      onClick={[Function]}
                      page={9}
                      selected={false}
                    >
                      <li>
                        <a
                          aria-current={null}
                          aria-label="Page 9"
                          onClick={[Function]}
                          onKeyPress={[Function]}
                          role="button"
                          tabIndex="0"
                        >
                          9
                        </a>
                      </li>
                    </PageView>
                    <PageView
                      activeClassName="selected"
                      activeLinkClassName="activePage"
                      key="9"
                      onClick={[Function]}
                      page={10}
                      selected={false}
                    >
                      <li>
                        <a
                          aria-current={null}
                          aria-label="Page 10"
                          onClick={[Function]}
                          onKeyPress={[Function]}
                          role="button"
                          tabIndex="0"
                        >
                          10
                        </a>
                      </li>
                    </PageView>
                    <li
                      className="next"
                    >
                      <a
                        className="next-icon"
                        onClick={[Function]}
                        onKeyPress={[Function]}
                        role="button"
                        tabIndex="0"
                      >
                        <span>
                          Next
                        </span>
                      </a>
                    </li>
                  </ul>
                </PaginationBoxView>
              </div>
            </ArticleContainer>
          </Connect(ArticleContainer)>
        </Route>
      </withRouter(Connect(ArticleContainer))>
    </Router>
  </BrowserRouter>
</Provider>
`;
